File: arch.py
import torch.nn as nn
import torch
from tqdm import tqdm
import copy
import numpy as np

class GFlowNet():
    def __init__(
            self,
            n_hidden_layers=2,
            hidden_dim=64,
            lr_model=1e-3,
            lr_logz=1e-1,
            device='cpu'
    ):
        """
        A GFlowNet class
        :param n_hidden_layers: Number of hidden layers
        :param hidden_dim: Number of hidden units
        :param lr_model: learning rate of the model
        :param lr_logz: learning rate of  logZ
        :param device: device for model and images
        """
        hidden_layers = nn.ModuleList()
        for i in range(n_hidden_layers):
            hidden_layers.append(nn.Linear(hidden_dim, hidden_dim))
            hidden_layers.append(nn.ELU())

        self.forward_model = nn.Sequential(
            nn.Linear(3, hidden_dim),
            nn.ELU(),
            *hidden_layers,
            nn.Linear(hidden_dim, 4)
        ).to(device)
        self.backward_model = copy.deepcopy(self.forward_model)
        self.logz = nn.Parameter(torch.tensor(0.0, device=device))
        self.device = device
        self.optimizer = torch.optim.Adam(
            [
                {'params': self.forward_model.parameters(), 'lr': lr_model},
                {'params': self.backward_model.parameters(), 'lr': lr_model},
                {'params': [self.logz], 'lr': lr_logz},
            ]
        )

    def init_state(self, env, batch_size):
        """
        Initializes s_0
        :param env: The environment to train in
        :param batch_size: Batch size
        :return: initialized state, tensor of shape (batch_size, 3) with each state consisting of [step,x,y]
        """
        x = torch.zeros(batch_size, 3, device=self.device)
        x[:,1:] += env.start
        return x

    @staticmethod
    def step(x, action):
        """
        Taking a step in the environment
        :param x: batch of states
        :param action: action to take
        :return: batch of new states
        """
        x_prime=x.clone()
        x_prime[:,1:] = x[:,1:] + action
        x_prime[:,0] = x[:,0] +1
        return x_prime

    @staticmethod
    def get_dist(policy, off_policy):
        """
        Get the distribution for the policy given the output of the network
        :param policy: output of the Neural Network
        :param off_policy: None to train on-policy. Otherwise a constant to add to sigma.
        :return: torch.dist.MultivariateNormal object
        """
        mus, vars = torch.tensor_split(policy, [2], dim=1)
        vars = nn.functional.sigmoid(vars)*0.9+0.1 # for mapping in valid [0.1-1] range
        policy_dist = torch.distributions.MultivariateNormal(mus, torch.diag_embed(vars))

        if not off_policy:
            return policy_dist, None


        exploration_dist = torch.distributions.MultivariateNormal(mus, torch.diag_embed(vars+off_policy))
        return policy_dist, exploration_dist

    def get_action(self, x, off_policy):
        """
        Get the actions for a given batch of states
        :param x: batch of states, Tensor of shape (batch_size, 3)
        :param off_policy: None to train on-policy. Otherwise a constant to add to sigma
        :return: actions, log_probs of the actions.
        Each is a Tensor of shape (batch_size, 2).
        For the actions: [step in x, step in y]
        """
        forward_policy = self.forward_model(x)
        policy_dist, exploration_dist = self.get_dist(forward_policy, off_policy)
        if off_policy:
            actions =exploration_dist.sample()
        else:
            actions = policy_dist.sample()
        log_probs = policy_dist.log_prob(actions)
        return actions, log_probs

    def get_backward_log_probs(self, states, actions):
        """
        Get the log_probs of the backward pass
        :param states: batch of states taken from the trajactories
        :param actions: batch of actions taken during forward pass
        :return:log_probs of the actions
        """
        backward_policy = self.backward_model(states)
        policy_dist,_ = self.get_dist(backward_policy, None)
        log_probs = policy_dist.log_prob(actions)
        return log_probs

    def train(self, env, batch_size=64, trajectory_length=2, n_iterations=5000, off_policy=None):
        """
        Train the model, Loss is Trajectory Balance Loss
        :param env: The environment to train in
        :param batch_size: Batch size
        :param trajectory_length: Fixed length of the trajectory
        :param n_iterations: number of iterations to train
        :param off_policy: None, constant or list
        None or 0 to train on-policy.
        A constant (int, float) will be added to the variance and lead to higher exploration.
        A list of int or float will be used as a schedule. The length must be equal to n_iterations.
        :return: tuple(List of losses, list of logZs, True Logz), last 4096 trajectories for visualization
        """

        losses = []
        trajectory_vis_n = 2048 # keep the last n trajectories for visualization
        trajectory_vis = torch.zeros((trajectory_vis_n, trajectory_length+1, 3), device=self.device)
        logzs = []
        logz_true = env.log_partition
        self.forward_model.train()
        self.backward_model.train()
        if not off_policy:
            exploration_schedule = [None] * n_iterations
        elif isinstance(off_policy, list):
            assert len(off_policy) == n_iterations, "Length of off_policy must be equal to n_iterations"
            exploration_schedule = off_policy
        else:
            assert isinstance(off_policy, int) or isinstance(off_policy, float), "no valid off_policy given"
            exploration_schedule = torch.linspace(off_policy, 0, n_iterations)

        progress_bar = tqdm(range(n_iterations), desc="Training...")
        for i in progress_bar:
            self.optimizer.zero_grad()
            x = self.init_state(env, batch_size)
            trajectory = torch.zeros(batch_size, trajectory_length+1, 3, device=self.device)
            log_probs_forward = torch.zeros((batch_size,), device=self.device)
            log_probs_backward = torch.zeros((batch_size,), device=self.device)

            for t in range(trajectory_length):
                actions, log_probs = self.get_action(x, exploration_schedule[i])
                log_probs_forward += log_probs
                x_prime = self.step(x, actions)
                trajectory[:,t+1,:] += x_prime
                x = x_prime

            for t in range(trajectory_length,1,-1):
                log_probs = self.get_backward_log_probs(trajectory[:,t,:], trajectory[:,t,1:]-trajectory[:,t-1,1:])
                log_probs_backward += log_probs

            log_reward = env.log_reward(trajectory[:,-1,1:])
            loss = torch.mean((self.logz+log_probs_forward-log_probs_backward-log_reward)**2)
            loss.backward()
            self.optimizer.step()
            losses.append(loss.item())
            logzs.append(self.logz.item())
            trajectory_vis = torch.cat((trajectory_vis, trajectory), dim=0)
            if len(trajectory_vis)>trajectory_vis_n:
                trajectory_vis = trajectory_vis[-trajectory_vis_n:]

            if i%20 == 0:
                progress_bar.set_postfix({
                    "Loss": f"{loss.item():.3f}",
                    "logZ": f"{self.logz.item():.3f}",
                    "logZ True": f"{logz_true.item():.3f}"
                })

        return (losses, logzs, logz_true.item()), trajectory_vis

    def inference(self, env, batch_size=4096, trajectory_length=2):
        """
        Sample from the model
        :param env: Environment to sample from
        :param batch_size: Number of trajectories to sample
        :param trajectory_length: Fixed length of the trajectory
        :return: Tensor of shape (batch_size, trajectory_length+1, 2) with the trajectorys of all samples
        """
        self.forward_model.eval()
        self.backward_model.eval()
        with torch.no_grad():
            trajectory = torch.zeros((batch_size, trajectory_length + 1, 3), device=self.device)
            trajectory[:, 0, 1:] += env.start
            x = self.init_state(env, batch_size)

            for t in range(trajectory_length):
                action, _ = self.get_action(x, None)
                x_prime = GFlowNet.step(x, action)
                trajectory[:,t+1,:] += x_prime
                x=x_prime

        return trajectory

File: front/src/arch.js

//getting the ml part in the frontend (work in progress)
export let tf = null;

if (window.tf) {
    tf = window.tf;
} else {
    window.addEventListener("load", () => {
        if (window.tf) {
            tf = window.tf;
            tf.setBackend("cpu");
        }
    });
}

export class MultivariateNormal{
    constructor(
        mus,
        sigmas
    ){
        /**
         * As tensorflow.js does not support Multivariate Normal distributions this is a helper class
         * It is limited to the case where the covariance matrix is defined by \sigma I,
         * e.g. no covariance and identical variance for all dimensions.
         * @param {tf.tensor} mus - A tensor of shape (batchsize,2) representing mu_x, mu_y.
         * @param {tf.tensor} sigmas - A tensor of shape(batchsize,1) representing sigma
         */

        this.mus = mus
        this.sigmas = sigmas
    }

    sample(){
        let samples_raw = tf.randomNormal(this.mus.shape, 0,1);
        let samples_scaled = samples_raw.mul(this.sigmas);
        let samples = samples_scaled.add(this.mus);
        return samples;
    }

    log_prob(samples) {
        // TODO needs checking
        console.log("lp 1")
        const d = this.mus.shape[1];  // Number of dimensions (should be 2 for 2D)
        const log2pi = tf.scalar(Math.log(2 * Math.PI));
        // Compute (x - mu) / sigma for each sample
        const diff = samples.sub(this.mus);
        const diff_scaled = diff.div(this.sigmas);
        // Compute the squared term (x - mu)^2 / sigma^2
        const squared_term = diff_scaled.square();
        // Sum over the dimensions (axis 1)
        const sum_squared = squared_term.sum(1);
        // Log probability calculation: log p(x) = -d/2 * log(2*pi) - sum(log(sigma)) - 1/2 * sum((x - mu)^2 / sigma^2)
        const log_sigma = tf.log(this.sigmas);
        const log_prob = tf.neg(
            tf.mul(
                tf.scalar(d / 2).mul(log2pi),
                tf.scalar(1)
            )
        ).sub(log_sigma.sum(1))
          .sub(sum_squared.mul(tf.scalar(0.5)));

        return log_prob;
    }



}


export class GFlowNet{
    constructor(
        n_hidden_layers=2,
        hidden_dim=64,
        lr_model=1e-3,
        lr_logz=1e-1,
    ){
        /**
         * A GFlowNet class.
         * @param {number} nHiddenLayers - Number of hidden layers.
         * @param {number} hiddenDim - Number of hidden units.
         * @param {number} lrModel - Learning rate of the model.
         * @param {number} lrLogZ - Learning rate of logZ.
         */
        tf.setBackend("cpu")
        this.forward_model = this.createModel(n_hidden_layers, hidden_dim);
        this.backward_model = this.createModel(n_hidden_layers, hidden_dim);
        this.logz = tf.variable(tf.scalar(0.0));
        this.optimizer_logz = tf.train.adam(lr_logz)
        this.optimizer_model = tf.train.adam(lr_model)
    };

    createModel(nHiddenLayers, hiddenDim) {
        const model = tf.sequential();
        model.add(tf.layers.dense({ inputShape: [3], units: hiddenDim, activation: 'elu' }));
        for (let i = 0; i < nHiddenLayers; i++) {
            model.add(tf.layers.dense({ units: hiddenDim, activation: 'elu' }));
        }
        model.add(tf.layers.dense({ units: 3 }));
        return model;
    }

    init_state(env, batch_size){
        /**
         * Initializes s_0
         * @param {Object} env - The environment to train in.
         * @param {number} batch_size - The batch size.
         * @returns {tf.Tensor} - Initialized state, a tensor of shape (batch_size, 3)
         *                        where each state consists of [step, x, y].
         */
        const x = tf.tidy(() =>{
            let a = tf.zeros([batch_size,1]); //state counter starts with 0
            let b = env.start.reshape([1,2]); // start state
            let c = tf.tile(b, [batch_size,1]);
            return tf.concat([a,c],1) //tensor of shape (batch_size, 3) with each state (step,x,y)
        });
        return x;
    }

    static step(x, action) {
        /**
         * Taking a step in the environment
         * @param {tf.Tensor} x - Batch of states (shape: [batch_size, 3])
         * @param {tf.Tensor} action - Action to take (shape: [batch_size, 2])
         * @return {tf.Tensor} - Batch of new states
         */
        return tf.tidy(() => {
            const stepIncrease = tf.onesLike(x.slice([0, 0], [-1, 1]));
            const newPos = x.slice([0, 1], [-1, 2]).add(action);
            return tf.concat([x.slice([0, 0], [-1, 1]).add(stepIncrease), newPos], 1);
        });



        /*
        let x_new = tf.tidy(() => {
            let updatedState = x.slice([0, 1], [-1, 2]).add(action); // x[:,1:] + action
            let updatedTime = x.slice([0, 0], [-1, 1]).add(1); // x[:,0] + 1

            return tf.concat([updatedTime, updatedState], 1); // Combine along axis 1
        });
        console.log("step taken: new x",x_new);
        return x_new; */
    }

    static get_dist (policy, off_policy){
        /**
         * Get the distribution for the policy given the output of the network
         * @param {tf.Tensor} policy - output of the Neural Network
         * @param {number} off_policy - A constant to add to sigma.
         * @returnm {[MultivariateDist, multivariateDist]} distributions - policy distribution and exploration distribution
         */
        const [mus, sigmas_raw] = tf.split(policy, [2,1], 1); // Split along axis 1 (columns)
        const sigmas = sigmas_raw.sigmoid().mul(0.9).add(0.1);
        let policy_dist = new MultivariateNormal(mus, sigmas);
        if(!off_policy) return [policy_dist, null];
        let exploration_dist = new MultivariateNormal(mus, sigmas.add(tf.scalar(off_policy)));
        return [policy_dist, exploration_dist];
    }

    get_action(x, off_policy){
        /**
         * Computes actions for a given batch of states.
         *
         * @param {tf.Tensor} x - A tensor of shape (batch_size, 3) representing the batch of states.
         * @param {number|null} [offPolicy=null] - A constant added to sigma for off-policy training.
         *                                         Set to null for on-policy training.
         * @returns {{ actions: tf.Tensor, logProbs: tf.Tensor }} - An object containing:
         *          - `actions`: A tensor of shape (batch_size, 2) representing [step in x, step in y].
         *          - `logProbs`: A tensor of shape (batch_size, 2) containing log probabilities of the actions.
         */
        const forward_policy = this.forward_model.apply(x);
        const [policy_dist, exploration_dist] = GFlowNet.get_dist(forward_policy, off_policy);
        const actions = off_policy ? exploration_dist.sample() : policy_dist.sample();
        const log_probs = policy_dist.log_prob(actions);

        return [actions, log_probs];
    }

    get_backward_log_probs(states, actions){
        /**
         * Get the log_probs of the backward pass
         * @param {tf.Tensor} states - batch of states taken from the trajactories
         * @param {tf.Tensor} actions - batch of actions taken during forward pass
         * @returns {tf.Tensor} log_probs of the actions
         */
        return tf.tidy(() => {
            const backward_policy = this.backward_model.apply(states);
            const [policy_dist, _] = GFlowNet.get_dist(backward_policy, null);
            return policy_dist.log_prob(actions);
        });
    }
    
    async train(
        env, 
        batch_size = 64, 
        trajectory_length = 2, 
        n_Iterations = 1024,
        off_policy = null, 
        lossFn = "Trajectory Balance"
    ) {
        const exploration_schedule = Array.isArray(off_policy) ? off_policy : tf.linspace(off_policy || 0, 0, n_Iterations).arraySync();
        console.log("exploration scheduled, start loop")
        for (let i = 0; i < n_Iterations; i++) {
            const logz_loss = this.logz.clone()
            let loss, non_logz_loss;
            this.optimizer_model.minimize(() => {

                return tf.tidy(() => {
                    const x = this.init_state(env, batch_size);
                    let log_probs_forward = tf.zeros([batch_size]);
                    let log_probs_backward = tf.zeros([batch_size]);
                    let trajectory = tf.zeros([batch_size, trajectory_length + 1, 3]);

                    for (let t = 0; t < trajectory_length; t++) {
                        const [actions, logProbs] = this.get_action(x, exploration_schedule[i]);
                        log_probs_forward = log_probs_forward.add(logProbs);
                        const x_prime = GFlowNet.step(x, actions);
                        trajectory = trajectory.slice([0, 0, 0], [batch_size, t + 1, 3]).concat(x_prime.expandDims(1), 1);
                    }
                    console.log("trajectory updated")

                    if (lossFn === "Trajectory Balance") {
                        for (let t = trajectory_length; t > 1; t--) {
                            const logProbs = this.get_backward_log_probs(trajectory.slice([0, t, 0], [batch_size, 1, 3]).reshape([batch_size, 3]),
                                trajectory.slice([0, t, 1], [batch_size, 1, 2]).reshape([batch_size, 2]).sub(
                                    trajectory.slice([0, t - 1, 1], [batch_size, 1, 2]).reshape([batch_size, 2])
                                )
                            );
                            log_probs_backward = log_probs_backward.add(logProbs);
                            console.log("logprobs calculated")
                        }
                    }

                    const logReward = env.log_reward(trajectory.slice([0, trajectory_length - 1, 1], [batch_size, 1, 2]).reshape([batch_size, 2]));
                    const non_logz_loss_temp = log_probs_forward.sub(log_probs_backward).sub(logReward);
                    non_logz_loss = tf.keep(non_logz_loss_temp.clone());
                    const loss_temp = tf.mean(non_logz_loss_temp.add(logz_loss).square());
                    loss = tf.keep(loss_temp.clone());
                    return loss_temp
                });
            });
            this.optimizer_logz.minimize(() =>{
                return tf.tidy(() => {
                    const loss_temp = tf.mean(non_logz_loss.add(this.logz).square());
                    non_logz_loss.dispose();
                    return loss_temp;
                });
            });
            loss.print();
            loss.dispose();
        }
    }

}